### ISSUE TITLE
twisted.names.server lacks test coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6887]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6887|
|Type|enhancement|
|Created|2013-12-30 13:30:59Z|
|Branch|https://github.com/twisted/twisted/tree/names-server-coverage-6887-2|

Add test coverage for twisted.names.server

Broken out from [#6700](https://github.com/twisted/twisted/issues/6700)

<details><summary>Searchable metadata</summary>

```
trac-id__6887 6887
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_names_server_coverage_6887_2 branches/names-server-coverage-6887-2
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1388410259000000 1388410259000000
changetime__1388758822000000 1388758822000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [41043]) Branching to 'names-server-coverage-6887'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj removed owner|
|-|-|

Ready for review in log:branches/names-server-coverage-6887
 * Created a new branch and merged in the test related changed from 6700
 * Addressed all exarkun's code review comments (ticket:6700#comment:6)

Forced a new build:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/names-server-coverage-6887
 * Some spurious twistedchecker warnings about emacs test headers (line too long) and comments
 * Some warnings about missing docstrings for really trivial test fakes - adding
   docstrings would be pointless and make the tests less readable I think.

Some replies to exarkuns code review comments below...

Replying to [ticket:6700#comment:6 exarkun]:
> Thanks.
>   1. twisted/names/test/test_server.py

>     1. Maybe `NoresponseDNSServerFactory` should be
>     `NoResponseDNSServerFactory`.

r41048

>     1. A `pass` statement following a docstring is superfluous - eg as in `AllowQueryArguments` and `NoresponseDNSServerFactory.sendReply`.

r41049

>     1. The API documentation for `NoresponseDNSServerFactory.allowQuery`'s
>     `args` parameter seems somewhat silly.  It *must* be positional arguments
>     because it is `*args`.  I don't think this is worth saying in a docstring.
>     Relatedly, though, it seems like it might be better to define these
>     methods with the correct signature anyway.

r41053.

>     1. All new "root" classes should be new-style (regardless of Python 3
>     considerations) unless they are helpers for testing behavior related to
>     classic classes (iow always subclass `object` if you're not subclassing
>     something else).  It may also make sense for non-root classes to be forced
>     to be new-style as well (mix `object` in if you're subclassing a classic
>     class) but this is slightly more complicated so I'm not going to make a
>     blanket statement about it.

r41056

>     1. No "simple suites" - eg, no `class DummyAuthority: pass`.

Not sure what you mean by this? In fact the only reason I'm using classes there
is to provide meaningful names in test failure messages eg "<class
DummyAuthority> is not equal to <class DummyCache>" instead of "<object 0x1> is
not equal to <object 0x2>".

>     1. I think the coding standard says to use `assertIs` instead of `assertIdentical` now.

r41058 (But I don't see it in the standard https://twistedmatrix.com/documents/current/core/development/policy/test-standard.html)

>     1. `test_cacheOverride` doesn't actually demonstrate the first element is
>     used as the cache (it could easily by the last element or the middle
>     element or a random element and the test would still pass).

r41059. (Wow. Well spotted!)

>     1. The `assertIsInstance` in `test_buildProtocolProtocolOverride` seems
>     redundant with the assertion made by
>     `test_buildProtocolDefaultProtocolType`.  Also I find that test methods
>     with only one assertion create a higher quality test suite than test
>     methods with multiple assertions.

r41061


>   1. The wrapping applied throughout seems to be too narrow.

r41063

Thanks again for the review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @wallrj|
|-|-|

> > No "simple suites" - eg, no class DummyAuthority: pass.
> Not sure what you mean by this? In fact the only reason I'm using classes there is to provide meaningful names in test failure messages eg "<class DummyAuthority> is not equal to <class DummyCache>" instead of "<object 0x1> is not equal to <object 0x2>".

Just write it the tall way:

```
class DummyAuthority:
    pass
```

instead of the short way:

```
class DummyAuthority: pass
```

The latter is what I call a "simple suite" - a ":" with no following "\n".  I think the Python grammar *used* to use this language but apparently it doesn't anymore (there seems to be no replacement vocabulary so I keep saying "simple suites").

> > I think the coding standard says to use assertIs instead of assertIdentical now.
> r41058 (But I don't see it in the standard  https://twistedmatrix.com/documents/current/core/development/policy/test-standard.html)

Hmm, indeed.  I see that the implementation defines `assertIs` and then makes `assertIdentical` an alias for it.  Maybe what the coding standard should say is to use the primary name for the assertion methods instead of any of the aliases.  I'll file a ticket maybe.

> > The assertIsInstance in test_buildProtocolProtocolOverride seems redundant with the assertion made by test_buildProtocolDefaultProtocolType.
> r41061

I think this adds too much vertical whitespace to `test_buildProtocolProtocolOverride`.  Perhaps the coding standard is misleading or ambiguous here - but the two/three blank line separator rules aren't meant to apply inside functions (class and function definitions inside functions should be rare enough that this doesn't much hurt readability and adding the extra vertical whitespace here reduces the separator effect of the top-level vertical whitespace).  Perhaps this needs to be clarify in the policy docs?

There are still some untested branches in `twisted.names.server`.  They all seem to have to do with logging behavior.  I'm not sure how you'd prefer to approach this.  On the one hand, the (existing) logging code is pretty lame.  On the other hand, it'd be pretty nice to have 100% coverage. :)

Perhaps a little bit of refactoring could help?  For example, each use of `if self.verbose: log.msg(...)` could be replaced by something like `self._verboseLog(...)` with `_verboseLog` containing the conditional.  Then a couple unit test for the behavior of `_verboseLog` with `self.verbose == 0` and `self.verbose == 1` would verify that it's going to do the right thing in each case.

I leave it entirely up to you whether you want to tackle this now or delay it to a later ticket.  If the latter, please do at least file the later ticket.

Buildbot looks good (as good as it ever does anyway).  Please address the above comments and merge if you're comfortable or re-submit if you'd like any more feedback (only the final point seems like it might require any).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

(In [41102]) Branching to 'names-server-coverage-6887-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set status to `assigned`|
|-|-|

Replying to [exarkun](#note_3):
> > > No "simple suites" - eg, no class DummyAuthority: pass.
> > Not sure what you mean by this? In fact the only reason I'm using classes there is to provide meaningful names in test failure messages eg "<class DummyAuthority> is not equal to <class DummyCache>" instead of "<object 0x1> is not equal to <object 0x2>".
> Just write it the tall way:

Done. r41101

> I think this adds too much vertical whitespace to `test_buildProtocolProtocolOverride`.  Perhaps the coding standard is misleading or ambiguous here - but the two/three blank line separator rules aren't meant to apply inside functions (class and function definitions inside functions should be rare enough that this doesn't much hurt readability and adding the extra vertical whitespace here reduces the separator effect of the top-level vertical whitespace).  Perhaps this needs to be clarify in the policy docs?

Done.

> There are still some untested branches in `twisted.names.server`.  They all seem to have to do with logging behavior.  I'm not sure how you'd prefer to approach this.  On the one hand, the (existing) logging code is pretty lame.  On the other hand, it'd be pretty nice to have 100% coverage. :)
> Perhaps a little bit of refactoring could help?  For example, each use of `if self.verbose: log.msg(...)` could be replaced by something like `self._verboseLog(...)` with `_verboseLog` containing the conditional.  Then a couple unit test for the behavior of `_verboseLog` with `self.verbose == 0` and `self.verbose == 1` would verify that it's going to do the right thing in each case.

I refactored most of the log lines as you suggested.

But I decided not to touch

 * https://twistedmatrix.com/trac/browser/branches/names-server-coverage-6887-2/twisted/names/server.py?rev=41108#L154
 * https://twistedmatrix.com/trac/browser/branches/names-server-coverage-6887-2/twisted/names/server.py?rev=41108#L454

...where it looked like building the log messages might be expensive and might
affect performance if done unconditionally.

> Buildbot looks good (as good as it ever does anyway).  Please address the above comments and merge if you're comfortable or re-submit if you'd like any more feedback (only the final point seems like it might require any).

I'll merge later if / when the builders go green.

> Thanks again!

Thankyou.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set status to `closed`|
|-|-|

(In [41109]) Merge names-server-coverage-6887-2

Author: rwall
Reviewer: exarkun
Fixes: [#6887](https://github.com/twisted/twisted/issues/6887)
Refs: [#6700](https://github.com/twisted/twisted/issues/6700)

Add test coverage for twisted.names.server
